config.org
#+TITLE: My emacs configuration

* Set user data

Place your private configuration here! Remember, you do not need to run 'doom
sync' after modifying this file!

Some functionality uses this to identify you, e.g. GPG configuration, email
clients, file templates and snippets.

#+begin_src emacs-lisp
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
(setq user-full-name "Zamachi"
      user-mail-address "dstefa.dimitrijevic7@gmail.com")
#+end_src

* Basic UI Config

** Font config

Doom exposes five (optional) variables for controlling fonts in Doom. Here
are the three important ones:

 + `doom-font'
 + `doom-variable-pitch-font'
 + `doom-big-font' -- used for `doom-big-font-mode' use this for
   presentations or streaming.

They all accept either a font-spec, font string ("Input Mono-12"), or xlfd
font string. You generally only need these two:
 - (setq doom-font (font-spec :family "monospace" :size 12 :weight 'semi-light)
        doom-variable-pitch-font (font-spec :family "sans" :size 13))

#+begin_src emacs-lisp

;;fontovi
(setq  doom-font(font-spec :family "FiraCode NF" :size 14)
       doom-variable-pitch-font(font-spec :family "FiraCode NF" :size 18))
#+end_src


** Color theme configuration

 There are two ways to load a theme. Both assume the theme is installed and
 available. You can either set `doom-theme' or manually load a theme with the
 `load-theme' function.
#+begin_src emacs-lisp
;;doom theme
(setq doom-theme 'doom-horizon)
#+end_src

** Number line mode
This determines the style of line numbers in effect. If set to `nil', line
numbers are disabled. For relative line numbers, set this to `relative'.

#+begin_src emacs-lisp
(setq display-line-numbers-type t)
(dolist (mode '(org-mode-hook
                term-mode-hook
                shell-mode-hook
                treemacs-mode
                eshell-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 0))))
#+end_src

* Keybinding configurations
** General

KORISTI SE DA SETUJE KEYBINDINGE U GLOBALU I ZA SPECIFICAN MODE(major & minor (?) )
keymaps specificira za koji mode(ili modove) vaze pravila. Mode se definise kao:
IME-MODA-map (npr. clojure-mode-map)
prefixi specificiraju kako se ulazi u definer, zatim se iz samog definera izvlace bindovi, global-prefix radi bez obzira gde se nalazili (I think?)

#+begin_src emacs-lisp
;; (use-package! general
;;   :config
;;   (general-create-definer zama/leader-keys
;;     :keymaps '(normal visual emacs)
;;     :global-prefix "SPC z")

;;   (zama/leader-keys
;;     "t"  '(:ignore t :which-key "toggles")
;;     "tt" '(counsel-load-theme :which-key "choose theme")))

#+end_src

[[https://github.com/noctuid/general.el][General dokumentacija]]
** Evil

Vim emulation sloj za emacs
Shift-V ulazi u visual(line) mode
Ctr-G(escape) ulazi u normal mode)
slovo i sluzi za insert
"dd" u normal modu brise liniju
"u" u normal modu radi undo (samo brisanja ?)
Ctrl-Z switchuje izmedju normal mode(Vim/evil) i "emacs" mode

[[https://github.com/emacs-evil/evil][Evil dokumentacija
]]
| Evil normal mode |                                    |
| Skracenica       | Opis                               |
| C-v              | duplira liniju                     |
| u                | undo                               |
| C-z              | switch na Emacs mode               |
| i                | Switch na insert mode              |
| Shift-V          | Visual line mode                   |
| dd               | brise liniju                       |
| d                | Brise selektovani sadrzaj i kopira |
| y                | Yank(copy) selektovanog            |
| P                | Nalepi sadrzaj PRE kursora         |
| p                | Nalepi sadrzaj NAKON kursora       |
| C-w              | Window manager (?)                 |
| j                | Next visual line                   |
| k                | Previous Visual line               |
| C-a              | Beginning of the line              |
| C-e              | End of the line                    |

| Evil insert mode |                  |
| Skracenica       | opis             |
| C-g/escape       | Switch na normal |
| Shift-insert     | Paste            |

| Evil visual line mode |                                  |
| Skracenica            | opis                             |
|-----------------------+----------------------------------|
| g c                   | zakomentarise selektovane linije |
** Evil collections

Dodatni keybindinzi koje sam Evil ne pokriva po difoltu

[[https://github.com/emacs-evil/evil-collection][Evil collection dokumentacija]]
* UI Configuration
** Better Modeline

Instalirati fontove ako se ne vide! *M-x install all font icons*

** Solaire mode

Na Doom Emacs-u instalirano po Default-u

** Which key

Konfigurisano po default-u na Doom Emacsu.

** Ivy and Counsel

[[https://oremacs.com/swiper/][Ivy]]je genericki completion alat ( koristi ivy-mode ).
Swiper je alternativa za ugradjeni isearch koji dobro radi sa Ivyjem
Counsel pruzi ekvivalente za Emacs komande(sa nekim dodatnim opcijama) (counsel-mode)

#+begin_src emacs-lisp
        ;;koristi ivy package, koji ima bolji autocomplete za meta- funkcional.(medju ostalim funkc.)

        ;; (use-package! ivy
        ;;   :diminish
        ;;   :bind (("C-s" . swiper)
        ;;          :map ivy-minibuffer-map
        ;;          ("TAB" . ivy-alt-done)
        ;;          ("C-f" . ivy-alt-done)
        ;;          ("C-l" . ivy-alt-done)
        ;;          ("C-j" . ivy-next-line)
        ;;          ("C-k" . ivy-previous-line)
        ;;          :map ivy-switch-buffer-map
        ;;          ("C-k" . ivy-previous-line)
        ;;          ("C-l" . ivy-done)
        ;;          ("C-d" . ivy-switch-buffer-kill)
        ;;          :map ivy-reverse-i-search-map
        ;;          ("C-k" . ivy-previous-line)
        ;;          ("C-d" . ivy-reverse-i-search-kill))
        ;;   :config
        ;;   (ivy-mode 1))
        ;;obogacuje ivy packet sa opisom funkcionalnosti i keybindovima(ako ih imaju)
        ;; (use-package! ivy-rich
        ;;   :after ivy
        ;;   :init
        ;;   (ivy-rich-mode 1))

;;         counsel paket, koji koristi ivy rich, da dodatno obogati meta-pretrage, C-x C-f i dr.
        (use-package! counsel
          :after ivy-rich
          :config
          (counsel-mode 1))
#+end_src

Kako se koristi

| funkcija          | precica | opis                                       |
|-------------------+---------+--------------------------------------------|
| Swiper            | SPC s b | Trazi string po regexu za dati buffer      |
| Find file         | SPC .   | Pronadji fajl                              |
| Counsel buffer    | SPC b B | Bira se na koji buffer ce se prebaciti     |
| Describe function | SPC h f | Opisuje funkciju                           |
| Describe variable | SPC h v | Opisuje promenljivu                        |
| Describe face     | SPC h F | Opisuje face atribut fontova               |
| Describe key      | SPC h k | Opisuje uneti keybind(ukoliko je bindovan) |
| Describe mode     | SPC h m | Opisuje trenutni (major?) mode             |
|                   |         |                                            |

Ako se u M-x u ne vidi opis funkcionalnosti, uraditi reset ivy-rich-mode-a

** Helpful

koristi counsel funkcionalnosti i helpful funkcionalnosti da overriduje defaultna ponasanja emacs-a npr. za describe-function ili bilo koje druge describe- funkcionalnosti
Na Doomu konfigurisano po default-u

** Text scaling
*** Hydra
Hydra se koristi kod buffer cycling-a, ali i za skaliranje teksta

Primer koriscenja: zamislimo da imamo C-c j i C-c k bindovano u konfiguraciji, i zelimo te dve komande da pozivamo u nekoj proizvoljnoj sekvenci, npr. C-c jjkkjkjkj, ovo nam dozvoljava Hydra paket, da ne pretiskamo C-c svaki put, vec samo jednom i da u sekvenci mozemo da zadajemo jjkkkjj naredbe, sto smo ovde iskoristili za povecavanje teksta.

#+begin_src emacs-lisp
  ;; (use-package! hydra
  ;;   :defer t)

  ;; (defhydra hydra-text-scale (:timeout 4)
  ;;   "scale text"
  ;;   ("j" text-scale-increase "in")
  ;;   ("k" text-scale-decrease "out")
  ;;   ("f" nil "finished" :exit t))

  ;; (zama/leader-keys
  ;;   "ts" '(hydra-text-scale/body :which-key "scale text"))
#+end_src

* Org mode
** Better font faces
#+begin_src emacs-lisp
(defun efs/org-font-setup ()
  ;; Replace list hyphen with dot
  (font-lock-add-keywords 'org-mode
                          '(("^ *\\([-]\\) "
                             (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "•"))))))

  ;; Set faces for heading levels
  (dolist (face '((org-level-1 . 1.2)
                  (org-level-2 . 1.1)
                  (org-level-3 . 1.05)
                  (org-level-4 . 1.0)
                  (org-level-5 . 1.1)
                  (org-level-6 . 1.1)
                  (org-level-7 . 1.1)
                  (org-level-8 . 1.1)))
    (set-face-attribute (car face) nil :font "FiraCode NF" :weight 'regular :height (cdr face)))

  ;; Ensure that anything that should be fixed-pitch in Org files appears that way
  (set-face-attribute 'org-block nil :foreground nil :inherit 'fixed-pitch)
  (set-face-attribute 'org-code nil   :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-table nil   :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-checkbox nil :inherit 'fixed-pitch))
#+end_src
** Basic org mode conf
#+begin_src emacs-lisp
  (defun efs/org-mode-setup ()
    (org-indent-mode)
    (variable-pitch-mode 1)
    (auto-fill-mode 0)
    (visual-line-mode 1)
    (setq evil-auto-indent nil))

  (use-package! org
    :commands (org-capture)
    :hook (org-mode . efs/org-mode-setup)
    :config
    (setq org-ellipsis " ▼"
          org-hide-emphasis-markers t)
    (efs/org-font-setup))
#+end_src
** Nicer bullets for headings
#+begin_src emacs-lisp
(use-package! org-bullets
  :after org
  :hook (org-mode . org-bullets-mode)
  :custom
  (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

#+end_src
** Center org buffers
#+begin_src emacs-lisp
(defun efs/org-mode-visual-fill ()
  (setq visual-fill-column-width 100
        visual-fill-column-center-text t)
  (visual-fill-column-mode 1))

(use-package! visual-fill-column
  :hook (org-mode . efs/org-mode-visual-fill))
#+end_src
** Configure babel languages
    Doom ih konfigurise za nas...
** Structure templates
#+begin_src emacs-lisp
    ;; (with-eval-after-load 'org
    ;;   (require 'org-tempo)

    ;;   (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
    ;;   (add-to-list 'org-structure-template-alist '("py" . "src python"))
    ;;   (add-to-list 'org-structure-template-alist '("lua" . "src lua"))

    ;;   )

(after! org
  (require 'org-tempo)
  (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
  (add-to-list 'org-structure-template-alist '("py" . "src python"))
  (add-to-list 'org-structure-template-alist '("lua" . "src lua"))
  )
#+end_src

** Auto-tangle configuration files

Basically automatically tangles this config file(saves it to another file), but just the codeblocks
  ;; (defun efs/org-babel-tangle-config ()
  ;;   (when (string-equal (buffer-file-name)
  ;;                       (expand-file-name "~/.doom.d/config.org"))
  ;;     ;; Dynamic scoping to the rescue
  ;;     (let ((org-confirm-babel-evaluate nil))
  ;;       (org-babel-tangle))))

  ;; (add-hook 'org-mode-hook (lambda () (add-hook 'after-save-hook #'efs/org-babel-tangle-config)))

No

* Development
** Commenting
Koristi se Evil nerd commenter, u Visual Line Mode(Shift-V) selektovati linije i pritisnuti *g c*

** Languages
*** Language Servers

LSP je Language Server Protocol koji se koristi da dostavi informaciju o autocompletion-u i slicnim stvarima vezanim za sintaksu jezika.

#+begin_src emacs-lisp
;;(add-hook! lsp-mode
(defun lsp-breadcrumb-setup()
           (setq lsp-headerline-breadcrumb-segments '(path-up-to-project file symbols))
           (lsp-headerline-breadcrumb-mode))

#+end_src

Usage of LSP - works only if language is provided:

| command                        | shortcut    | description                                       |
|--------------------------------+-------------+---------------------------------------------------|
| Find references                | C-c l g r   | Pronalazi sve reference selektovanog simbola radi |
| Find definitions               | C-c l g g   | Isto kao iznad, samo pronalazi  definicije        |
| Rename                         | C-c l r r   | Preimenuje selektovan simbol                      |
| Flymake diagnostics buffer     |             | Dijagnostika gresaka u datom fajlu - greske, warn |
| Code formatting                | C-c l = =   | Reformatira kod kako ih language serv. formatira  |
| lsp-treemacs-symbols           | M-x command | Show a tree view of symbols in the file           |
| lsp-treemacs-references        | M-x command | Show a tree view for refs of symbol under cursor  |
| lsp-treemacs-error-list        | M-x command | Show a tree veiw for diagnostic msgs              |
| lsp-ivy-workspace-symb         | M-x command | Search for a symbol name in the current project   |
| lsp-ivy-global-workspace-sym   | M-x command | Searches in all active project workspaces         |
| treemacs                       |             | Opens the treemacs view on the side               |
| lsp-ui-sideline-enable         |             | Enables lsp-ui sideline (possibly won't work ?)   |
| lsp-headerline-breadcrumb-mode |             | Enables breadcrumb for the project                |
**** Documentation

[[https://github.com/emacs-lsp/lsp-ui][LSP-UI documentation]]
*** DAP mode and Debugging
Dap-mode je paket za debugging putem Debugging Adapter Protocol. Debager za svaki jezik se konfigurise posebno
[[https://github.com/emacs-lsp/dap-mode][Dokumentacija DAP-mode.]]


#+begin_src emacs-lisp
    ;; (require 'dap-node)
    ;; (dap-node-setup);; Automatically installs Node debug adapter if needed

    ;; Sa "C-c l d" otvori se dap-hydra prozor za lakse debagovanje
    ;; (general-define-key
    ;;   :keymaps 'lsp-mode-map
    ;;   :prefix lsp-keymap-prefix
    ;;   "d" '(dap-hydra t :wk "debugger"))
#+end_src

Ukljucuje se sa *M-x Dap-debug-mode*, a onda ce pitati da se selektuje konfiguracioni templejt. LSP-mode mora biti aktivan u bufferu. Konfiguracije se moze menjati sa *M-x dap-debug-recent.*  Editovanje konfiguracija se vrsi sa *M-x dap-debug-edit-templates.*
Konfiguracioni templejti se najcesce cuvaju u posebnom fajlu, unutar korenog direktorijuma projekta npr.,  tipa *debug.el* gde se nalepi templejt sledeceg izgleda:

(dap-register-debug-template
"ime-templejta"
(list :type "tip-debuggera"(npr node)
        :cwd nil <- podesava trenutni radni direktorijum
        :request "launch"
        :program nil <- koji program ce se pokrenuti
        :name "opet-ime")), koristi se konfiguracija sa *VSCodea*

Templejt treba evaluirati (*C-x C-e*) pre nego sto postane dostupan.

[[https://emacs-lsp.github.io/dap-mode/page/configuration/#dap-mode-configuration][Konfiguracija DAP mode-a.]]

**** Breakpoints

Breakpoint se dodaje sa *M-x dap-breakpoint-toggle* naredbom.
Da se otvori prikaz sa Breakpointima, pokrece se komanda *M-x dap-ui-breakpoints* i kada se zavrsi, moze se izaci pritiskom na taster q
*M-x dap-breakpoint-log-message* dodaje custom poruku koja se ispisuje kada debugger dodje na breakpoint. Sa *{ promenljiva iz koda }* mozemo evaluirati neki izraz.
Moze se dodati breakpoint i sa uslovom: *M-x dap-breakpoint-condition*. Uslov se pise u jeziku koji se debaguje (?)

Takodje mozemo dodati *watch expressions* - odnosno promenljive cije vrednosti zelimo da se ispisuju u panelu "Expressions", jednostavno pozovemo naredbu *M-x dap-ui-expressions-add ime-promenljive* i ta promenljiva iz koda ce se ispisivati u pogledu.
Uklanja se sa *M-x dap-ui-expressions-remove*.

Da se prikazu lokalne varijable, koristi se funkcija *M-x dap-ui-locals*.

**** Navigacija

Sa *M-x dap-next* se prelazi na sledecu liniju u kodu.

Promena funkcije na stacku se vrsi sa *dap-switch-stack-frame*

**** Sesije

Sesija se resetuje sa *M-x dap-debug-restart*.
Debugger se gasi sa *M-x dap-disconnect*
Prikaz svih sesija za debagovanje *M-x dap-ui-sessions* , ako se onda prebacimo na taj view i pretisnemo *D(u Emacs mode-u - C-z)* onda se obrise sesija.

Paleta komandi: *M-x dap-hydra*

**** REPL

	Aktivira se sa *M-x dap-ui-repl* ukoliko dap adapter za dati jezik podrzava interaktivnu konzolu. U sustini interaktivna konzola nam dozvoljava da tokom izvrsavanja menjamo kod - mogu se cak i vrednosti menjati dinamicki

****  VSCode debug config

		VSCode cuva konfiguracije debuggera u *.vscode folderu( debug.json )* i to se moze iskoristiti i u Emacsu, jer je isti debugger adapter i debug protokol!

*** TypeScript
    Configured by default by Doom ( ? not confirmed )
*** Lua

**** Love minor mode

Lua and love minor mode active by default

*** Python
	*Data konfiguracije je pisana za Doom Emacs! Moze se prilagoditi iza Spacemacs distribuciju, kao i za vanilla Emacs*
	
    Kada se konfigurise Emacs kao Python IDE, potrebno je preduzeti sledece korake:
        - Instalirati Pajton (3.3+ verzije) interpreter i sve potrebne dependency-je( automatski na Linux platformama ), *OBAVEZNO* instalirati *python3-venv* - neophodno da kreiramo virtuelno okruzenje za svaki projekat ( na Windowsu dolazi distribuirano uz sam Pajton ).
        - Instalirati *Pip* da bismo mogli instalirati dodatne pakete - i to pip3! *sudo apt-get install python3-pip* na Ubuntu( na Windowsu je distribuirano zajedno sa Pajtonom )
        - Napraviti *virtual environment* pomocu *venv* sa naredbom =python3 -m venv /path/to/new/virtual/environment=, gde je glavni direktorijum virtualnih okruzenja *$HOME(~) u .venvs folderu* , po defaultu( promeniti u konfiguraciji ispod ako je druga putanja u pitanju ). Na Winowsu koristiti samo python bez trojke. 
        - Aktivirati virtuelno okruzenje sa =source ~/.venvs/ime-okruzenja/bin/activate=: na Linux sistemima. Na Windowsu pokrenuti =./venvs/ime-okruzenja/Scripts/activate.ps1= u Powershellu.
          - Kada smo u virtuelnom okruzenju, potrebno je instalirati naredbom =pip3 install ime-paketa= :
            - *isort* - da soritra importe
            - *Pyflakes* - Proverava greske u fajlu
            - Language server za LSP protocol, 3 najzastupljenija:
	      1. *python-lsp-server* - instalirati sa =pip3 install python-lsp-server=.
	      2. *mspyls* - pratiti instrukcije na [[https://develop.spacemacs.org/layers/+lang/python/README.html#microsoft-python-language-server][sledecem linku.]] *PAZNJA!* mozda je potrebno, pored ovog Emacs plugin-a instalirati i globalno sam Microsoft Python server! Dodatno, ukoliko ne koristite Doom ili Spacemacs, verovatno cete morati da ispratite i sledece korake da [[https://emacs-lsp.github.io/lsp-python-ms/ ][ukljucite LSP plugin za Emacs]].
	      3. *Pyright* - Microsoftov Python LS pisan u TypeScriptu. Pre svega je potrebno instalirati sam Pyright server, za to su nam neophodni *npm* i *node*. Kada njih instaliramo, pozivom komande =npm install -g pyright= cemo na globalnom nivou instalirati server, a zatim za svaki projekat mozemo odabrati da li cemo ga koristiti ili ne (putem *.dir-local.el* fajla ili nekim drugim putem). Dodatno, ukoliko ne koristite Doom ili Spacemacs, verovatno cete morati da ispratite i sledece korake da [[https://emacs-lsp.github.io/lsp-pyright/][ukljucite LSP plugin za Emacs]].
	    - Dodatne potrebne biblioteke za projekat - npr. *Pillow*
	    * Ako zelimo, mozemo i instalirati posebnu verziju Python interpretera koju ce projekat koristiti, za ovo koristimo *Pyenv* koji se dodatno konfigurise...

Doom automatski prepozna sta je projekat( ako ima *.git* folder npr. ), dodatna konfiguracija najverovatnije nece biti potrebna, sve sto treba je da izaberemo virtuelno okruzenje koje ce koristiti projekat sa =pyvenv-= naredbama. Ukoliko je potrebno, projekat se moze konfigurisati sa *.dir-local.el* fajlom, na sledeci nacin:
    
#+begin_src emacs-lisp :tangle no
 ((nil . ((pyvenv-workon . "ime-venva")))
 	(python-mode . ((python-backend . lsp);;NOTE works only for Spacemacs
 			(python-lsp-server . pyright))));;NOTE works only for Spacemacs
#+end_src

Izmeniti ime venva, i ime servera(ili backenda ako sek oristi anaconda) za sosptvene projekte.

    Na kraju je samo potrebno konfigurisati *pyvenv* paket koji upravlja Python *venv* okruzenjima( promeniti *WORKON_HOME* ako je putanja direktorijuma gde cuvamo okruzenja drugacija  ).

#+begin_src emacs-lisp
(use-package! pyvenv
  :init
  (setenv "WORKON_HOME" "~/.venvs/")
  :config
  ;; (pyvenv-mode t)

  ;; Set correct Python interpreter
  (setq pyvenv-post-activate-hooks
        (list (lambda ()

                (if (eq system-type 'windows-nt )
                    (setq python-shell-interpreter (concat pyvenv-virtual-env "Scripts/python"))
                  (setq python-shell-interpreter (concat pyvenv-virtual-env "bin/python"))))))
  (setq pyvenv-post-deactivate-hooks
        (list (lambda ()
                (if (eq system-type 'windows-nt )
                    (setq python-shell-interpreter "python")
                  (setq python-shell-interpreter "python3"))))))
#+end_src


    Projekat treba da izgleda na sledeci nacin:

fooproject
|-- .git
|-- src
     |-- foo.py
|-- .dir-locals.el
|-- pyrightconfig.json

*pyrightconfig.json* fajl je opcioni, ali takodje sluzi za konfiguraciju, samo sto se njime konfigurise *Pyright* server:

#+begin_src json :tangle no
{
  "include": [
    "src"
  ],
  "executionEnvironments": [
    {
      "root": "src"
    }
  ]
}
#+end_src
*PRIMER KONFIGURACIJE*:
Ovime smo konkretno specificirali gde sve Pyright treba da se pokrene, dali smo src kao nivo celog foldera i svim *.py* fajlovima u njemu>
Jos opcija za *pyrightconfig.json* na sledecem [[https://github.com/microsoft/pyright/blob/main/docs/configuration.md][linku]].

PAZNJA: Projekat se umesto ugradjene *venv* funkcije moze konfigurisati i naprednijim *virtualenv* paketom, cak i sa *virtualenvwrapper* koji je napredniji od prethodnog, za ovo pogledati sledeci video: https://www.youtube.com/watch?v=r-BHx7VNX5s (NAPOMENA: u videu se projekat konfigurise za Python2! Ispratiti kako se konfigurise za Python3 i da li treba dodatno nastelovati *virtualenvwrapper* da radi u terminalu preko .bashrc i PATH varijabli).

***  C++

#+begin_src emacs-lisp
  ;; (require 'dap-cpptools)
  ;; (add-hook 'c++-mode-hook 'lsp)
#+end_src


** Company mode

Provides a nicer in-buffer completion interface.
Company box further enhances the looks and icons and presentation
Aktivno sve po defaultu od strane Doom-a
** Projectile

 Emacs ima koncept DIRECTORY LOCAL VARIABLES - kreira se fajl u repo folderu koji konfigurise emacs varijable, koje ce se primeniti na odredjene fajlove u projektu svaki put kad se otvori bafer koji cita fajl iz datog foldera:
  	- y - otvara .dir-locals.el unutar (trenutnog ?) repo-a, koji predstavlja taj konfiguracioni fajl u sustini...
	NAREDBE KOJE SE MOGU KORISTITI:
		+ projectile-project-run-cmd - koja naredba se izvrsava kada se uradi C-c p-p

 Kada se zavrsi editovanje i sacuva fajl, treba da se uradi evaluate (M-:) sa naredbom: (hack-dir-local-variables) i reloaduje ove promenljive iz dir-locals.el fajla.

#+begin_src emacs-lisp
  ;;projectile project interaction library for emacs. Offers functionalities for projects

  (use-package! counsel-projectile
    :after projectile
    :config (counsel-projectile-mode))
#+end_src

| funkcija                   | skracenica | opis                                         |
|----------------------------+------------+----------------------------------------------|
| projectile-edit-dir-locals | C-c p E    | Otvara .dir-locals.el unutar trenutnog repoa |
| projectile-project-run-cmd | C-c p-p    | Pokrece se projekat                          |
| hack-dir-local-variables   |            | Refreshuje .dir-locals.el kada se izmeni     |
| projectile menu            | C-c p      | Otvara projectile menu                       |


| Projectile menu komande   |            |                                                    |
| Funkcija                  | skracenica | opis                                               |
| projectile-switch-project | p          | Menja projekat                                     |
| fuzzy search              | f          | Fuzzy search u projektnom folderu                  |
| Testing project           | P          | Pokrece testove u datom projektu (?)               |
| (?)                       | Alt-o      | Prikazuje dodatne funkcionalnosti za dati projekat |
| counsel-projectile-rg     |            | Grep komanda, samo sto trazi po celom projektu     |

** Basic Magit setup

Doom setupuje Magit korektno

| Funkcija       | skracenica | opis                                                    |
|----------------+------------+---------------------------------------------------------|
| Magit status   | C-x g      | Status verzionnisanja                                   |
| Staging        | C-x g s    | Staging fajlova                                         |

config.el
;;; $DOOMDIR/config.el -*- lexical-binding: t; -*-
(setq user-full-name "Zamachi"
      user-mail-address "dstefa.dimitrijevic7@gmail.com")

;;fontovi
(setq  doom-font(font-spec :family "FiraCode NF" :size 14)
       doom-variable-pitch-font(font-spec :family "FiraCode NF" :size 18))

;;doom theme
(setq doom-theme 'doom-horizon)

(setq display-line-numbers-type t)
(dolist (mode '(org-mode-hook
                term-mode-hook
                shell-mode-hook
                treemacs-mode
                eshell-mode-hook))
  (add-hook mode (lambda () (display-line-numbers-mode 0))))

;; (use-package! general
;;   :config
;;   (general-create-definer zama/leader-keys
;;     :keymaps '(normal visual emacs)
;;     :global-prefix "SPC z")

;;   (zama/leader-keys
;;     "t"  '(:ignore t :which-key "toggles")
;;     "tt" '(counsel-load-theme :which-key "choose theme")))

;;koristi ivy package, koji ima bolji autocomplete za meta- funkcional.(medju ostalim funkc.)

        ;; (use-package! ivy
        ;;   :diminish
        ;;   :bind (("C-s" . swiper)
        ;;          :map ivy-minibuffer-map
        ;;          ("TAB" . ivy-alt-done)
        ;;          ("C-f" . ivy-alt-done)
        ;;          ("C-l" . ivy-alt-done)
        ;;          ("C-j" . ivy-next-line)
        ;;          ("C-k" . ivy-previous-line)
        ;;          :map ivy-switch-buffer-map
        ;;          ("C-k" . ivy-previous-line)
        ;;          ("C-l" . ivy-done)
        ;;          ("C-d" . ivy-switch-buffer-kill)
        ;;          :map ivy-reverse-i-search-map
        ;;          ("C-k" . ivy-previous-line)
        ;;          ("C-d" . ivy-reverse-i-search-kill))
        ;;   :config
        ;;   (ivy-mode 1))
        ;;obogacuje ivy packet sa opisom funkcionalnosti i keybindovima(ako ih imaju)
        ;; (use-package! ivy-rich
        ;;   :after ivy
        ;;   :init
        ;;   (ivy-rich-mode 1))

;;         counsel paket, koji koristi ivy rich, da dodatno obogati meta-pretrage, C-x C-f i dr.
        (use-package! counsel
          :after ivy-rich
          :config
          (counsel-mode 1))

;; (use-package! hydra
;;   :defer t)

;; (defhydra hydra-text-scale (:timeout 4)
;;   "scale text"
;;   ("j" text-scale-increase "in")
;;   ("k" text-scale-decrease "out")
;;   ("f" nil "finished" :exit t))

;; (zama/leader-keys
;;   "ts" '(hydra-text-scale/body :which-key "scale text"))

(defun efs/org-font-setup ()
  ;; Replace list hyphen with dot
  (font-lock-add-keywords 'org-mode
                          '(("^ *\\([-]\\) "
                             (0 (prog1 () (compose-region (match-beginning 1) (match-end 1) "•"))))))

  ;; Set faces for heading levels
  (dolist (face '((org-level-1 . 1.2)
                  (org-level-2 . 1.1)
                  (org-level-3 . 1.05)
                  (org-level-4 . 1.0)
                  (org-level-5 . 1.1)
                  (org-level-6 . 1.1)
                  (org-level-7 . 1.1)
                  (org-level-8 . 1.1)))
    (set-face-attribute (car face) nil :font "FiraCode NF" :weight 'regular :height (cdr face)))

  ;; Ensure that anything that should be fixed-pitch in Org files appears that way
  (set-face-attribute 'org-block nil :foreground nil :inherit 'fixed-pitch)
  (set-face-attribute 'org-code nil   :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-table nil   :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-verbatim nil :inherit '(shadow fixed-pitch))
  (set-face-attribute 'org-special-keyword nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-meta-line nil :inherit '(font-lock-comment-face fixed-pitch))
  (set-face-attribute 'org-checkbox nil :inherit 'fixed-pitch))

(defun efs/org-mode-setup ()
  (org-indent-mode)
  (variable-pitch-mode 1)
  (auto-fill-mode 0)
  (visual-line-mode 1)
  (setq evil-auto-indent nil))

(use-package! org
  :commands (org-capture)
  :hook (org-mode . efs/org-mode-setup)
  :config
  (setq org-ellipsis " ▼"
        org-hide-emphasis-markers t)
  (efs/org-font-setup))

(use-package! org-bullets
  :after org
  :hook (org-mode . org-bullets-mode)
  :custom
  (org-bullets-bullet-list '("◉" "○" "●" "○" "●" "○" "●")))

(defun efs/org-mode-visual-fill ()
  (setq visual-fill-column-width 100
        visual-fill-column-center-text t)
  (visual-fill-column-mode 1))

(use-package! visual-fill-column
  :hook (org-mode . efs/org-mode-visual-fill))

;; (with-eval-after-load 'org
    ;;   (require 'org-tempo)

    ;;   (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
    ;;   (add-to-list 'org-structure-template-alist '("py" . "src python"))
    ;;   (add-to-list 'org-structure-template-alist '("lua" . "src lua"))

    ;;   )

(after! org
  (require 'org-tempo)
  (add-to-list 'org-structure-template-alist '("el" . "src emacs-lisp"))
  (add-to-list 'org-structure-template-alist '("py" . "src python"))
  (add-to-list 'org-structure-template-alist '("lua" . "src lua"))
  )

;;(add-hook! lsp-mode
(defun lsp-breadcrumb-setup()
           (setq lsp-headerline-breadcrumb-segments '(path-up-to-project file symbols))
           (lsp-headerline-breadcrumb-mode))

;; (require 'dap-node)
;; (dap-node-setup);; Automatically installs Node debug adapter if needed

;; Sa "C-c l d" otvori se dap-hydra prozor za lakse debagovanje
;; (general-define-key
;;   :keymaps 'lsp-mode-map
;;   :prefix lsp-keymap-prefix
;;   "d" '(dap-hydra t :wk "debugger"))

(use-package! pyvenv
  :init
  (setenv "WORKON_HOME" "~/.venvs/")
  :config
  ;; (pyvenv-mode t)

  ;; Set correct Python interpreter
  (setq pyvenv-post-activate-hooks
        (list (lambda ()

                (if (eq system-type 'windows-nt )
                    (setq python-shell-interpreter (concat pyvenv-virtual-env "Scripts/python"))
                  (setq python-shell-interpreter (concat pyvenv-virtual-env "bin/python"))))))
  (setq pyvenv-post-deactivate-hooks
        (list (lambda ()
                (if (eq system-type 'windows-nt )
                    (setq python-shell-interpreter "python")
                  (setq python-shell-interpreter "python3"))))))

;; (require 'dap-cpptools)
;; (add-hook 'c++-mode-hook 'lsp)

;;projectile project interaction library for emacs. Offers functionalities for projects

(use-package! counsel-projectile
  :after projectile
  :config (counsel-projectile-mode))

(add-hook! lsp-mode
           '(rainbow-delimiters-mode lsp-breadcrumb-setup))

(use-package! devdocs
  :commands(devdocs-install devdocs-search devdocs-lookup))

;;(add-hook 'python-mode-hook
;;          (lambda () (setq-local devdocs-current-docs '("python~3.9"))))

(if (eq system-type 'windows-nt )
    (use-package! powershell
      :commands shell
      :config
      ;; Change default compile command for powershell
      (add-hook 'powershell-mode-hook
                (lambda ()
                  (set (make-local-variable 'compile-command)
                       (format "powershell.exe -NoLogo -NonInteractive -Command \"& '%s'\"" ))))))

(use-package! dired
  :commands(dired dired-jump)
  :custom ((dired-listing-switches "-agho --group-directories-first"))
  :config
  (evil-collection-define-key 'normal 'dired-mode-map
    "h" 'dired-single-up-directory
    "l" 'dired-single-buffer))

(use-package! dired-single
  :commands (dired dired-jump))

;; (use-package! dired-open
;;   :commands (dired dired-jump)
;;   :config
;;   (setq dired-open-extensions '(("png" . "feh")
;;                                 ("mkv" . "mpv"))))

;; (setq gc-cons-treshold (* 2 1000 1000))
